src/rpc/network/socket_communicator.cc

```diff
@@ -255,65 +288,122 @@ void SocketReceiver::Finalize() {
+  while (static_cast<size_t>(received_bytes) < sizeof(int64_t)) {
+    int64_t max_len = sizeof(int64_t) - received_bytes;
+    int64_t tmp = socket->Receive(
+      reinterpret_cast<char*>(&data_size) + received_bytes,
+      max_len);
+    if (tmp == -1) {
+      if (received_bytes > 0) {
+        // We want to finish reading full data_size
+        continue;
+      }
+      return -1;
     }
-    if (data_size < 0) {
-      LOG(FATAL) << "Recv data error (data_size: " << data_size << ")";
-    } else if (data_size == 0) {
-      // This is an end-signal sent by client
+    received_bytes += tmp;
+  }
+  return data_size;
+}
+
+void RecvData(TCPSocket* socket, char* buffer, const int64_t &data_size,
+  int64_t *received_bytes) {
+  while (*received_bytes < data_size) {
+    int64_t max_len = data_size - *received_bytes;
+    int64_t tmp = socket->Receive(buffer + *received_bytes, max_len);
+    if (tmp == -1) {
+      // Socket not ready, not more data to read
       return;
-    } else {
-      char* buffer = nullptr;
-      try {
-        buffer = new char[data_size];
-      } catch(const std::bad_alloc&) {
-        LOG(FATAL) << "Cannot allocate enough memory for message, "
-                   << "(message size: " << data_size << ")";
+    }
+    *received_bytes += tmp;
+  }
+}
+
+void SocketReceiver::RecvLoop(
+  std::unordered_map<int /* Sender (virtual) ID */,
+    std::shared_ptr<TCPSocket>> sockets,
+  std::unordered_map<int /* Sender (virtual) ID */,
+    std::shared_ptr<MessageQueue>> queues,
+  runtime::Semaphore *queue_sem) {
+  std::unordered_map<int, std::unique_ptr<RecvContext>> recv_contexts;
+  SocketPool socket_pool;
+  for (auto& socket : sockets) {
+    auto &sender_id = socket.first;
+    socket_pool.AddSocket(socket.second, sender_id);
+    recv_contexts[sender_id] = std::unique_ptr<RecvContext>(new RecvContext());
+  }
+
+  // Main loop to receive messages
+  for (;;) {
+    int sender_id;
+    // Get active socket using epoll
+    std::shared_ptr<TCPSocket> socket = socket_pool.GetActiveSocket(&sender_id);
+    if (queues[sender_id]->EmptyAndNoMoreAdd()) {
+      // This sender has already stopped
+      if (socket_pool.RemoveSocket(socket) == 0) {
+        return;
       }
-      received_bytes = 0;
-      while (received_bytes < data_size) {
-        int64_t max_len = data_size - received_bytes;
-        int64_t tmp = socket->Receive(buffer+received_bytes, max_len);
-        CHECK_NE(tmp, -1);
-        received_bytes += tmp;
+      continue;
+    }
+
+    // Nonblocking socket might be interrupted at any point. So we need to
+    // store the partially received data
+    std::unique_ptr<RecvContext> &ctx = recv_contexts[sender_id];
+    int64_t &data_size = ctx->data_size;
+    int64_t &received_bytes = ctx->received_bytes;
+    char*& buffer = ctx->buffer;
+
+    if (data_size == -1) {
+      // This is a new message, so receive the data size first
+      data_size = RecvDataSize(socket.get());
```

Comment:
can we always receive the complete data size?

Reply:
yes, this RecvDataSize method is blocking and will always receive full 8 bytes.